Introduce builder pattern for server initialization

jcdyer opened the pull request:
I'm working on a fork of raft-rs, where I'm hoping to implement some missing raft features including persistent logs, snapshotting, and cluster reconfiguration.  

As I was getting started, I noticed that a lot of the configuration was hardcoded, and thought I'd give a shot at making it more extensible by introducing a ServerBuilder.  This way, new features can be added without constantly changing the public signature for creation.

I also tried to disentangle the server creation and event loop creation.  The envisioned interface is demoed in the example code.

Let me know if this is a direction you're interested in pursuing, or if there are any modifications you'd like to see.

Hoverbear commented:
Hey! :)

Thanks for submitting this!

I really like the builder idea. I have to admit I'm not a huge fan of builders which `.expect()` if they are missing required arguments. How do you feel about making it so the initial `.builder()` taking the required arguments, and the builder methods being used for optionals?

Disentangling the Raft server from the event loop is a goal I'd like to pursue as well. @skade and I have been discussing a way to implement Raft using futures and then building a way to attach it to a Tokio core.

arthurprs commented:
I can help reviewing if you want to send more PRs this way.

Hoverbear commented:
Thanks, @arthurprs !

jcdyer commented:
> How do you feel about making it so the initial .builder() taking the required arguments, and the builder methods being used for optionals?

I like that idea.  It leaves less guesswork to the caller, and places more correctness on the type system.  

The current implementation has the interface you're asking for (I believe), but still uses `Option`s internally, because I'm not sure how to handle passing the ownership of the non-clonable StateMachine otherwise.  This lets me pass it over without having to create an object to replace it, but the expect should now be irrefutable.  I made the finalize method consume the builder, so it can't get reused erroneously.

Hoverbear commented:
So what you can do is actually make `finalize` consume `self` instead of just borrow it:

```rust
pub struct ServerBuilder<L, M>
where
    L: Log,
    M: StateMachine,
{
    id: ServerId,
    addr: SocketAddr,
    peers: HashMap<ServerId, SocketAddr>,
    store: L,
    state_machine: M,
    max_connections: usize,
    election_min_millis: u64,
    election_max_millis: u64,
    heartbeat_millis: u64,
}

impl <L, M> ServerBuilder<L, M>
where
    L: Log,
    M: StateMachine
{
    fn new(id: ServerId, addr: SocketAddr, peers: HashMap<ServerId, SocketAddr>, store: L, state_machine: M) -> ServerBuilder<L, M> {
        /// Create a ServerBuilder with default values
        /// for optional members.
        ServerBuilder {
            id: id,
            addr: addr,
            peers: peers,
            store: store,
            state_machine: state_machine,
            max_connections: 128,
            election_min_millis: 150,
            election_max_millis: 350,
            heartbeat_millis: 60,
        }
    }

    pub fn finalize(self) -> Result<Server<L, M>> {
        Server::create(
            self.id,
            self.addr,
            self.peers,
            self.store,
            self.state_machine,
            self.election_min_millis,
            self.election_max_millis,
            self.heartbeat_millis,
            self.max_connections,
        )
    }
```

At this point, due to how the server works when you `.run()`, the need for a separate builder seems rather low. Since `Server` is, in a sense, finalized by using `.run()`. Maybe we could put the builder methods directly on `Server` instead?

jcdyer commented:
I took out the need for an extra .finalize() call, by adding a `ServerBuilder::run()` that calls back into `Server::run()`.  I kept the separate builder, though, for a few reasons:

* The state it contains is represented differently on the Server struct.
* Keeping the builder better captures the semantics of "configure, then run."  Configuration methods shouldn't be called on a running server, and this makes that impossible.
* It keeps the code more cleanly organized, as configuration code is all in one place, and server behavior in another.

arthurprs commented:
I wouldn't block the PR over this but I think only the absolutely required show be in new. So maybe peers should be optional in `with_peers`.

Hoverbear commented on src/server.rs:
```diff
@@ -76,39 +150,60 @@ pub struct Server<L, M>
     where L: Log,
           M: StateMachine
 {
+    pub fn new(
```
The fact that both a `create` and `new` exists confuses me. Is there a reason for this?

jcdyer replied:
Yes, but I think the naming could use some work.  I renamed `builder()` to `new()`, and the old `new()` to `create()`.  Note that `create()` is a non-public function.  It's just there to finalize the creation.  Maybe since `Server::new()` delegates to `ServerBuilder::new()`, I should make `ServerBuilder::finalize()` delegate to `Server::finalize()`, instead of `Server::create()`.  For the moment, that's all internal code, but it may become public if the run phase gets extracted, so it probably makes sense to give some thought to naming.

Hoverbear commented on src/server.rs:
```diff
@@ -137,16 +247,11 @@ impl<L, M> Server<L, M>
-        let actions = server.consensus.init();
-        server.execute_actions(&mut event_loop, actions);
-        event_loop.run(&mut server).map_err(From::from)
+    pub fn run(&mut self) -> Result<()> {
```
I like this 🍾 ! A better idea definitely.

Hoverbear commented on examples/hashmap.rs:
```diff
@@ -158,7 +158,12 @@ fn server(args: &Args) {
     // You probably shouldn't `.unwrap()` in production code unless you're totally sure it works
     // 100% of the time, all the time.
-    Server::run(id, addr, peers, persistent_log, state_machine).unwrap();
+    Server::new(id, addr, peers, log, state_machine)
```
I much prefer this style of an API over `Server::run` :)

jcdyer replied:
Thanks!

Hoverbear commented:
I would say I generally agree with @arthurprs .

Otherwise, the interaction/interplay between `ServerBuilder` and `Server` feels a bit awkward and weird to me... But that's partially due to the code being already a bit of a mess. I'm OK with merging this under the intention of cleaning things up more as we go. :)

Let me know if we're good for merge or there is something you'd like to touch up/address.

jcdyer commented:
I'm happy to clean up the peering set up before we merge. I'll give it a try tonight. I'm thinking either an "add_peer(ServerId, SocketAddr)" or an "add_peers(Iterator<(ServerId, SocketAddr)>)" (or IntoIterator<..>)

jcdyer commented on src/server.rs:
```diff
@@ -166,10 +271,12 @@ impl<L, M> Server<L, M>
+            .spawn(move || {
+                let mut server = try!(Server::create(id, addr, peers, store, state_machine, 1500, 3000, 1000, 129));
+                server.run()
+            })
```
This part feels like a wart.  I hoped that `spawn()` could be equivalent to `run()`, but `Server` can't be moved into the new thread.  

It may make sense to remove this method altogether, and let application code handle thread creation (so callers can pass in whatever configuration they desire), but I didn't want to take too much of an axe to the code in my first PR. :)

Hoverbear replied:
No problem. :) If you'd like to do structural changes I'm quite fine reviewing/accepting them. The code needs a lot of love and I haven't really been giving it. :(

jcdyer commented:
Ok.  I think this is good to go.  I added a with_peers method, though it only takes a HashMap for now.  I think I'll be able to make it work with an IntoIterator, but I need to figure out a little more about how to use the relevant traits.  That can come in another PR.  I also renamed Server::create to Server::finalize to match what the builder is doing.

jcdyer commented:
If you give it a thumbs up, I'll squash the commits into one before you merge.

Hoverbear commented:
I squashed them. :)

jcdyer commented:
![Image of Keytar Bear](https://wrongsideofthecamera.files.wordpress.com/2014/05/boston-old-state-house-keytar-bear-3.jpg)

Awesome!

Hoverbear commented:
![tell-me-more](https://cloud.githubusercontent.com/assets/130903/26104179/eb6d3e0e-3a3c-11e7-9c8a-f10f1f7c4581.jpeg)

jcdyer commented:
[Keytar Bear](http://keytarbear.com/): The finest street musician in Boston.
